### PR TITLE
[dv/kmac] Initial KMAC testplan

### PR DESCRIPTION
**Pull request by udinator**
Entropy interface hasn't been finalized yet so this is not included in the testplan yet.
A followup PR will be created to add entropy tests once ready.

There is no need to review this until the testplan meeting, just adding everyone to this PR for accessibility.

Signed-off-by: Udi Jonnalagadda <udij@google.com>

### CODE REVIEW

File: hw/ip/kmac/data/kmac_base_testplan.hjson [hunk 1]
```diff
@@ -5,27 +5,129 @@
...
+      tests: ["{name}_kdf"]
+    }
+    {
+      name: kdf_with_sw_access
+      desc: '''
+            Initiate a KDF operation as before, but this time also provide a "SW" message and key,
+            and try to also initiate the KMAC with the CMD register.
+            KMAC should ignore the "SW" commands/data and proceed with KDF.
+            After KDF is done, trigger a software-initiated KMAC process using the previously
+            available message/key.
```
Comment (eunchan) [inline hw/ip/kmac/data/kmac_base_testplan.hjson:100]: This also could be added to the error code. I will note that and add it later into the design. Thanks

**Review by eunchan — COMMENTED** [review]
LGTM.

File: hw/ip/kmac/data/kmac_base_testplan.hjson [hunk 2]
```diff
@@ -5,27 +5,129 @@
...
-            - TBD
-
-            **Checks**:
-            - TBD
+            KMAC smoke test will contain a number of rounds, and acts as a base for many other tests.
+            In each round, we run a full KMAC hashing operation:
+              - Randomly enable interrupts, operation mode (SHA3/SHAKE/CSHAKE/KMAC), key length (if
+                applicable), constraining all settings to be legal.
+              - Set function name to empty string and set customization string between 0-8 bits in
+                length if applicable.
```
[the remark this answers is not in the document]
Comment (udinator) [reply hw/ip/kmac/data/kmac_base_testplan.hjson:19]: Oh, ok! That definitely simplifies things a little bit. 
I'll update the test to have this be included then.
If this is the case, shall I still include the cshake test vectors? They all use the empty string function name.

[the remark this answers is not in the document]
Comment (eunchan) [reply hw/ip/kmac/data/kmac_base_testplan.hjson:19]: No. I think we don't need to test cSHAKE, cSHAKE in KMAC is assumed to support  KMAC operation only. Later we may see some lower coverage due to this, but I think we can safely waive them as we don't plan to use.

[the remark this answers is not in the document]
Comment (udinator) [reply hw/ip/kmac/data/kmac_base_testplan.hjson]: Sounds good! I will get rid of the cshake test vectors test then.

File: hw/ip/kmac/data/kmac_base_testplan.hjson [hunk 3]
```diff
@@ -5,27 +5,129 @@
...
-
-            **Checks**:
-            - TBD
+            KMAC smoke test will contain a number of rounds, and acts as a base for many other tests.
+            In each round, we run a full KMAC hashing operation:
+              - Randomly enable interrupts, operation mode (SHA3/SHAKE/CSHAKE/KMAC), key length (if
+                applicable), constraining all settings to be legal.
+              - Set function name to "KMAC" and set customization string as empty if applicable.
+              - Randomly set endianness of input msg and internal keccak state.
+              - Randomly provide a sideloaded key, do not set cfg.sideload.
```
Comment (udinator) [inline hw/ip/kmac/data/kmac_base_testplan.hjson:20]: randomly update configuration during absorption process

File: hw/ip/kmac/data/kmac_base_testplan.hjson [hunk 4]
```diff
@@ -5,27 +5,129 @@
...
+    {
+      name: error
+      desc: '''
+            Try several error sequences:
+
+            - Update key or prefix during absorption or process stage.
+            - Incorrect SHA3 control flow (e.g. write cmd.done after message absorption).
+            - Incorrect KMAC configurations (e.g. set KMAC strength as 512).
+
+            // TODO: Not all of these errors have been supported yet, and more errors might be
```
Comment (udinator) [inline hw/ip/kmac/data/kmac_base_testplan.hjson:114]: try setting function name to not "KMAC" encoded, should raise error

File: hw/ip/kmac/data/kmac_base_testplan.hjson [hunk 5]
```diff
@@ -5,27 +5,129 @@
...
+            '''
+      milestone: V2
+      tests: ["{name}_kdf_with_sw_key"]
+    }
+    {
+      name: error
+      desc: '''
+            Try several error sequences:
+
+            - Update key or prefix during absorption or process stage.
```
Comment (udinator) [inline hw/ip/kmac/data/kmac_base_testplan.hjson:110]: write more message/configuration during squeezing stage

File: hw/ip/kmac/data/kmac_base_testplan.hjson [hunk 6]
```diff
@@ -5,27 +5,129 @@
...
+      desc: '''
+            Initiate a KDF operation as before, but this time also provide a "SW" message and key,
+            and try to also initiate the KMAC with the CMD register.
+            KMAC should ignore the "SW" commands/data and proceed with KDF.
+            After KDF is done, trigger a software-initiated KMAC process using the previously
+            available message/key.
+            '''
+      milestone: V2
+      tests: ["{name}_kdf_with_sw_key"]
+    }
```
Comment (udinator) [inline hw/ip/kmac/data/kmac_base_testplan.hjson:104]: add a throughput test for sha3/shake/kmac
- sha3 doesn't need entropy

File: hw/ip/kmac/data/kmac_base_testplan.hjson [hunk 7]
```diff
@@ -5,27 +5,129 @@
...
       tests: ["{name}_smoke"]
     }
     {
-      name: feature1
-      desc: '''Add more test entries here like above.'''
-      milestone: V1
-      tests: []
+      name: long_message_and_output
+      desc: '''
+            Same as the smoke test, except with a message of up to 20KB.
```
Comment (udinator) [inline hw/ip/kmac/data/kmac_base_testplan.hjson:40]: max firmware size can be 200KB

Comment (udinator) [reply hw/ip/kmac/data/kmac_base_testplan.hjson:40]: average can be around 100KB

File: hw/ip/kmac/data/kmac_base_testplan.hjson (hunk 3 above)
Comment (udinator) [reply hw/ip/kmac/data/kmac_base_testplan.hjson:20]: moved this to the error test

**Comment by udinator** [comment]
@sriyerg @cindychip @weicaiyang could you give this a quick look so we can get this merged?

File: hw/ip/kmac/data/kmac_base_testplan.hjson [hunk 8]
```diff
@@ -5,27 +5,150 @@
...
+
+            - Update key/prefix/config during absorption/process/squeeze stage.
+            - Write msg to msg_fifo during process/squeeze stage
+            - When in KMAC mode, set the function name to not "KMAC"
+            - Incorrect SHA3 control flow:
+              - Issue Process/Run/Done cmds before issuing Start
+              - Issue Run/Done before issuing Process
+              - Issue Start after issuing Process
+              - If squeezing data, issue Start/Process after issuing Run
+              -
```
Comment (cindychip) [inline hw/ip/kmac/data/kmac_base_testplan.hjson:123]: Remove this `-`?

File: hw/ip/kmac/doc/checklist.md [hunk 9]
```diff
@@ -129,8 +129,8 @@ Coverage      | [SIM_COVERAGE_MODEL_ADDED][]          | Done        |
 Code Quality  | [TB_LINT_SETUP][]                     | Done        |
 Integration   | [PRE_VERIFIED_SUB_MODULES_V1][]       | Done        |
 Review        | [DESIGN_SPEC_REVIEWED][]              | Done        |
-Review        | [DV_PLAN_TESTPLAN_REVIEWED][]         | Not Started |
-Review        | [STD_TEST_CATEGORIES_PLANNED][]       | Not Started |
+Review        | [DV_PLAN_TESTPLAN_REVIEWED][]         | Done        |
+Review        | [STD_TEST_CATEGORIES_PLANNED][]       | Done        |
```
Comment (cindychip) [inline hw/ip/kmac/doc/checklist.md:133]: I saw most other IPs have exceptions here, such as power, security, and debug, I wonder if that also applies for KMAC?

File: hw/ip/kmac/data/kmac_base_testplan.hjson (hunk 8 above)
Comment (udinator) [reply hw/ip/kmac/data/kmac_base_testplan.hjson:123]: good catch, fixed.

File: hw/ip/kmac/doc/checklist.md (hunk 9 above)
Comment (udinator) [reply hw/ip/kmac/doc/checklist.md:133]: ah good point! you're right, there are no plans to do these testing in the UVM DV environment, will update accordingly.

File: hw/ip/kmac/doc/checklist.md [hunk 10]
```diff
@@ -129,8 +129,9 @@ Coverage      | [SIM_COVERAGE_MODEL_ADDED][]          | Done        |
 Code Quality  | [TB_LINT_SETUP][]                     | Done        |
 Integration   | [PRE_VERIFIED_SUB_MODULES_V1][]       | Done        |
 Review        | [DESIGN_SPEC_REVIEWED][]              | Done        |
-Review        | [DV_PLAN_TESTPLAN_REVIEWED][]         | Not Started |
```
Comment (weicaiyang) [inline hw/ip/kmac/doc/checklist.md:132]: TB_GEN_AUTOMATED should be N/A 
FPV_MAIN_ASSERTIONS_PROVEN seems only applicable for FPV, it should be also N/A. @cindychip LMK if it's not right

**Review by weicaiyang — APPROVED** [review]
LGTM with a small comment

File: hw/ip/kmac/doc/checklist.md (hunk 10 above)
Comment (cindychip) [reply hw/ip/kmac/doc/checklist.md:132]: yup, agree with weicai

Comment (udinator) [reply hw/ip/kmac/doc/checklist.md:132]: Hmm ok. I set TB_GEN_AUTOMATED to Done because of multiple parameter configs being verified, but I can change both to N/A as well. thanks for the comment!

**Review by cindychip — APPROVED** [review]

**Comment by udinator** [comment]
latest force push to resolve merge conflicts
